tfjs-layers/src/layers/nlp/tokenizers.ts

```diff
@@ -134,24 +136,338 @@ export abstract class Tokenizer extends Layer {
+
+    const wordsStr = await tensorArrTo2DArr<string>(words);
+
+    // Get all word pairs.
+    const first = wordsStr.map(arr => arr.slice(0, -1));
+    const second = wordsStr.map(arr => arr.slice(1, arr.length));
+
+    // Mask empty.
+    const nonEmptyMask = second.map(arr => arr.length > 0);
+    mask = mask.map((a, idx) => a && nonEmptyMask[idx]);
+    if (!mask.some(e => e)) {
+      return [words, mask];
+    }
+    const nonEmptyIndices = mask
+      .map((bool, idx) => bool ? idx : -1)
+      .filter(e => e !== -1);
+
+    const filteredFirst = nonEmptyIndices.map(idx => first[idx]);
+    const filteredSecond = nonEmptyIndices.map(idx => second[idx]);
+
+    // Get byte pair ranking in merge rules.
+    const pairs: string[][] = filteredFirst.map((firstSubArr, idx) => {
+      const secondSubArr = filteredSecond[idx];
+
+      return firstSubArr.map(
+        (char, idx) => [char, secondSubArr[idx]].join(' '));
+    });
+    const pairRanksTensor = await this.mergeRanks.lookup(
+      pairs.map(arr => tensor(arr)));
+    const pairRanks = await tensorArrTo2DArr<number>(pairRanksTensor);
+
+    // Get BPE pair ranks.
+    const minPairRank = pairRanks.map(arr => Math.min(...arr));
+    const pairFoundMask = minPairRank.map(
+      rank => rank !== this.mergeRanksLookupDefault);
+
+    // Tokens that cannot be further merged are marked as finished.
+    nonEmptyIndices.forEach((index, idx) => {
+      const update = pairFoundMask[idx];
+      mask[index] = update;
+    });
+    if (!mask.some(e => e)) {
+      return [words, mask];
+    }
+
+    function argMin(arr: number[]): number {
+      return arr.indexOf(Math.min(...arr));
+    }
+
+    const maskedPairRanks = pairRanks.filter((_, idx) => pairFoundMask[idx]);
+    const minPairRankIndices = maskedPairRanks.map(arr => argMin(arr));
+
+    // Get words and pairs to process.
+    const unfinishedWords = wordsStr.filter((_, idx) => mask[idx]);
+
+    const pairLeft = unfinishedWords.map(
+      (word, idx) => word[minPairRankIndices[idx]]);
+
+    const pairRight = unfinishedWords.map(
+      (word, idx) => word[minPairRankIndices[idx] + 1]);
+
+    const mergedPairs = pairLeft.map((left, idx) => {
+      const right = pairRight[idx];
+      return [left, right].join('');
+    });
+    const unfinishedWordsIndices = mask
+      .map((_, idx) => idx)
+      .filter((_, idx) => mask[idx]);
+
+    const mergedPairIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx]]);
+    const emptyStringIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx] + 1]);
+
+    mergedPairIndices.forEach((indices, idx) => {
+      const [wordIdx, charIdx] = indices;
+      const mergedPair = mergedPairs[idx];
+      wordsStr[wordIdx][charIdx] = mergedPair;
+    });
+
+    emptyStringIndices.forEach((indices) => {
+      const [wordIdx, charIdx] = indices;
+      wordsStr[wordIdx][charIdx] = '';
+    });
```

Comment:
Nit: Prefer for loops over `forEach`. `for (const [idx, indices] of mergedPairIndices.entries())`